File: js/src/lib/query-commands.js
/**
 * Query and control command handlers.
 *
 * Every option in this module addresses an execution that already exists in
 * the tracking store instead of launching a new one: --status, --list,
 * --upload-log, --stop, --terminate, --attach, --resume, --resume-all and
 * --cleanup. `dispatchQueryCommand` is the single entry point the CLI calls
 * before it starts interpreting the command line as a command to run.
 */

const { queryStatus, listExecutions } = require('./status-formatter');
const { ControlAction, controlExecution } = require('./execution-control');
const { attachExecution } = require('./execution-attach');
const { resumeAllExecutions, resumeExecution } = require('./execution-resume');
const { uploadExecutionLog } = require('./log-uploader');

/**
 * Print a handler result, returning the process exit code.
 * @param {{success: boolean, output?: string, error?: string, exitCode?: number}} result
 * @returns {number} Exit code
 */
function reportResult(result) {
  if (result.success) {
    if (result.output) {
      console.log(result.output);
    }
    return result.exitCode || 0;
  }
  console.error(`Error: ${result.error}`);
  return result.exitCode || 1;
}

function handleStatusQuery(store, identifier, outputFormat) {
  return reportResult(queryStatus(store, identifier, outputFormat));
}

function handleListQuery(store, outputFormat, options = {}) {
  return reportResult(listExecutions(store, outputFormat, options));
}

function handleUploadLogQuery(store, identifier) {
  const result = uploadExecutionLog(store, identifier);
  if (result.success) {
    return result.exitCode || 0;
  }
  console.error(`Error: ${result.error}`);
  return result.exitCode || 1;
}

function handleControlQuery(store, identifier, action) {
  return reportResult(controlExecution(store, identifier, action));
}

function handleAttachQuery(store, identifier, readOnly) {
  return reportResult(attachExecution(store, identifier, { readOnly }));
}

async function handleResumeQuery(store, identifier, command, outputFormat) {
  return reportResult(
    await resumeExecution(store, identifier, { command, outputFormat })
  );
}

function handleResumeAllQuery(store, outputFormat) {
  return reportResult(resumeAllExecutions(store, { outputFormat }));
}

/**
 * Handle --cleanup / --cleanup-dry-run.
 * Cleans up stale "executing" records (processes that crashed or were killed).
 * @param {?object} store - Execution store
 * @param {boolean} dryRun - If true, just report what would be cleaned
 * @returns {number} Exit code
 */
function handleCleanup(store, dryRun) {
  if (!store) {
    console.error('Error: Execution tracking is disabled.');
    return 1;
  }

  const result = store.cleanupStale({ dryRun });

  if (result.errors.length > 0) {
    for (const error of result.errors) {
      console.error(`Error: ${error}`);
    }
  }

  if (result.records.length === 0) {
    console.log('No stale records found.');
    return 0;
  }

  if (dryRun) {
    console.log(
      `Found ${result.records.length} stale record(s) that would be cleaned up:\n`
    );
  } else {
    console.log(`Cleaned up ${result.cleaned} stale record(s):\n`);
  }

  for (const record of result.records) {
    const startTime = new Date(record.startTime).toLocaleString();
    console.log(`  UUID: ${record.uuid}`);
    console.log(`  Command: ${record.command}`);
    console.log(`  Started: ${startTime}`);
    console.log(`  PID: ${record.pid || 'N/A'}`);
    console.log('');
  }

  if (dryRun) {
    console.log('Run with --cleanup to actually clean up these records.');
  }
  return 0;
}

function hasValue(value) {
  return value !== null && value !== undefined;
}

/**
 * Report whether any query/control mode is active.
 * @param {object} options - Parsed wrapper options
 * @returns {boolean} True when the invocation addresses a stored execution
 */
function hasQueryMode(options) {
  return (
    hasValue(options.status) ||
    options.list === true ||
    hasValue(options.uploadLog) ||
    hasValue(options.stop) ||
    hasValue(options.terminate) ||
    hasValue(options.attach) ||
    hasValue(options.resume) ||
    options.resumeAll === true ||
    options.cleanup === true
  );
}

/**
 * Run the query/control mode selected by the wrapper options, if any.
 * @param {object} options - Parsed wrapper options
 * @param {object} context - {store, command}
 * @returns {Promise<?number>} Exit code, or null when no query mode is active
 */
async function dispatchQueryCommand(options, context = {}) {
  const store = context.store || null;
  const outputFormat = options.outputFormat;

  if (hasValue(options.status)) {
    return handleStatusQuery(store, options.status, outputFormat);
  }
  if (options.list) {
    return handleListQuery(store, outputFormat, { running: options.running });
  }
  if (hasValue(options.uploadLog)) {
    return handleUploadLogQuery(store, options.uploadLog);
  }
  if (hasValue(options.stop)) {
    return handleControlQuery(store, options.stop, ControlAction.STOP);
  }
  if (hasValue(options.terminate)) {
    return handleControlQuery(
      store,
      options.terminate,
      ControlAction.TERMINATE
    );
  }
  if (hasValue(options.attach)) {
    return handleAttachQuery(store, options.attach, options.readOnly === true);
  }
  if (hasValue(options.resume)) {
    return await handleResumeQuery(
      store,
      options.resume,
      context.command || null,
      outputFormat
    );
  }
  if (options.resumeAll) {
    return handleResumeAllQuery(store, outputFormat);
  }
  if (options.cleanup) {
    return handleCleanup(store, options.cleanupDryRun);
  }

  return null;
}

module.exports = {
  dispatchQueryCommand,
  hasQueryMode,
  handleAttachQuery,
  handleCleanup,
  handleControlQuery,
  handleListQuery,
  handleResumeAllQuery,
  handleResumeQuery,
  handleStatusQuery,
  handleUploadLogQuery,
  reportResult,
};

File: rust/src/lib/query_commands.rs
//! Query and control command handlers.
//!
//! Every option in this module addresses an execution that already exists in
//! the tracking store instead of launching a new one: `--status`, `--list`,
//! `--upload-log`, `--stop`, `--terminate`, `--attach`, `--resume`,
//! `--resume-all` and `--cleanup`. `dispatch_query_command` is the single entry
//! point the CLI calls before it starts interpreting the command line as a
//! command to run (issue #162).

use crate::args_parser::WrapperOptions;
use crate::execution_attach::{attach_execution, ExecutionAttachResult};
use crate::execution_control::{control_execution, ControlAction, ExecutionControlResult};
use crate::execution_resume::{resume_execution, ExecutionResumeResult};
use crate::execution_resume_all::{resume_all_executions, ExecutionResumeAllResult};
use crate::execution_store::{CleanupOptions, ExecutionStore};
use crate::log_uploader::upload_execution_log;
use crate::status_formatter::{list_executions_filtered, query_status, StatusQueryResult};

/// Uniform shape every query handler reports back.
pub struct QueryOutcome {
    pub success: bool,
    pub output: Option<String>,
    pub error: Option<String>,
    pub exit_code: Option<i32>,
}

impl From<StatusQueryResult> for QueryOutcome {
    fn from(result: StatusQueryResult) -> Self {
        QueryOutcome {
            success: result.success,
            output: result.output,
            error: result.error,
            exit_code: None,
        }
    }
}

impl From<ExecutionControlResult> for QueryOutcome {
    fn from(result: ExecutionControlResult) -> Self {
        QueryOutcome {
            success: result.success,
            output: result.output,
            error: result.error,
            exit_code: None,
        }
    }
}

impl From<ExecutionAttachResult> for QueryOutcome {
    fn from(result: ExecutionAttachResult) -> Self {
        QueryOutcome {
            success: result.success,
            output: result.output,
            error: result.error,
            exit_code: result.exit_code,
        }
    }
}

impl From<ExecutionResumeResult> for QueryOutcome {
    fn from(result: ExecutionResumeResult) -> Self {
        QueryOutcome {
            success: result.success,
            output: result.output,
            error: result.error,
            exit_code: None,
        }
    }
}

impl From<ExecutionResumeAllResult> for QueryOutcome {
    fn from(result: ExecutionResumeAllResult) -> Self {
        QueryOutcome {
            success: result.success,
            output: result.output,
            error: result.error,
            exit_code: None,
        }
    }
}

/// Print a handler result, returning the process exit code.
pub fn report_result(result: QueryOutcome) -> i32 {
    if result.success {
        if let Some(output) = result.output {
            println!("{}", output);
        }
        return result.exit_code.unwrap_or(0);
    }
    if let Some(error) = result.error {
        eprintln!("Error: {}", error);
    }
    result.exit_code.unwrap_or(1)
}

/// Handle `--upload-log`, which reports its own exit code.
pub fn handle_upload_log_query(store: Option<&ExecutionStore>, identifier: &str) -> i32 {
    match upload_execution_log(store, identifier) {
        Ok(code) => code,
        Err(error) => {
            eprintln!("Error: {}", error);
            1
        }
    }
}

/// Handle `--cleanup` / `--cleanup-dry-run`.
///
/// Cleans up stale "executing" records (processes that crashed or were killed).
pub fn handle_cleanup(store: Option<&ExecutionStore>, dry_run: bool) -> i32 {
    let Some(store) = store else {
        eprintln!("Error: Execution tracking is disabled.");
        return 1;
    };

    let result = store.cleanup_stale(CleanupOptions {
        dry_run,
        ..Default::default()
    });

    for error in &result.errors {
        eprintln!("Error: {}", error);
    }

    if result.records.is_empty() {
        println!("No stale records found.");
        return 0;
    }

    if dry_run {
        println!(
            "Found {} stale record(s) that would be cleaned up:\n",
            result.records.len()
        );
    } else {
        println!("Cleaned up {} stale record(s):\n", result.cleaned);
    }

    for record in &result.records {
        let start_time_display = chrono::DateTime::parse_from_rfc3339(&record.start_time)
            .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string())
            .unwrap_or_else(|_| record.start_time.clone());

        println!("  UUID: {}", record.uuid);
        println!("  Command: {}", record.command);
        println!("  Started: {}", start_time_display);
        println!(
            "  PID: {}",
            record
                .pid
                .map(|p| p.to_string())
                .unwrap_or("N/A".to_string())
        );
        println!();
    }

    if dry_run {
        println!("Run with --cleanup to actually clean up these records.");
    }
    0
}

/// Run the query/control mode selected by the wrapper options, if any.
///
/// Returns the process exit code, or `None` when no query mode is active and
/// the invocation should be treated as a command to run.
pub fn dispatch_query_command(
    options: &WrapperOptions,
    store: Option<&ExecutionStore>,
    command: &str,
) -> Option<i32> {
    let output_format = options.output_format.as_deref();

    if let Some(ref identifier) = options.status {
        return Some(report_result(
            query_status(store, identifier, output_format).into(),
        ));
    }
    if options.list {
        return Some(report_result(
            list_executions_filtered(store, output_format, options.running).into(),
        ));
    }
    if let Some(ref identifier) = options.upload_log {
        return Some(handle_upload_log_query(store, identifier));
    }
    if let Some(ref identifier) = options.stop {
        return Some(report_result(
            control_execution(store, identifier, ControlAction::Stop).into(),
        ));
    }
    if let Some(ref identifier) = options.terminate {
        return Some(report_result(
            control_execution(store, identifier, ControlAction::Terminate).into(),
        ));
    }
    if let Some(ref identifier) = options.attach {
        return Some(report_result(
            attach_execution(store, identifier, options.read_only).into(),
        ));
    }
    if let Some(ref identifier) = options.resume {
        let new_command = if command.trim().is_empty() {
            None
        } else {
            Some(command)
        };
        return Some(report_result(
            resume_execution(store, identifier, new_command, output_format).into(),
        ));
    }
    if options.resume_all {
        return Some(report_result(
            resume_all_executions(store, output_format).into(),
        ));
    }
    if options.cleanup {
        return Some(handle_cleanup(store, options.cleanup_dry_run));
    }

    None
}
